staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -466,60 +456,58 @@ type processorListener struct {
 
 func newProcessListener(handler ResourceEventHandler, requestedResyncPeriod, resyncPeriod time.Duration, now time.Time) *processorListener {
 	ret := &processorListener{
-		pendingNotifications:  []interface{}{},
 		nextCh:                make(chan interface{}),
+		addCh:                 make(chan interface{}),
 		handler:               handler,
 		requestedResyncPeriod: requestedResyncPeriod,
 		resyncPeriod:          resyncPeriod,
 	}
 
-	ret.cond.L = &ret.lock
-
 	ret.determineNextResync(now)
 
 	return ret
 }
 
-func (p *processorListener) add(notification interface{}) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	p.pendingNotifications = append(p.pendingNotifications, notification)
-	p.cond.Broadcast()
+func (p *processorListener) add(stopCh <-chan struct{}, notification interface{}) {
+	select {
+	case <-stopCh:
+	case p.addCh <- notification:
+	}
 }
 
 func (p *processorListener) pop(stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 
+	// pendingNotifications is an unbounded slice that holds all notifications not yet distributed
+	// there is one per listener, but a failing/stalled listener will have infinite pendingNotifications
+	// added until we OOM.
+	// TODO This is no worse that before, since reflectors were backed by unbounded DeltaFIFOs, but
+	// we should try to do something better
+	var pendingNotifications []interface{}
+	var nextCh chan<- interface{}
+	var notification interface{}
 	for {
-		blockingGet := func() (interface{}, bool) {
-			p.lock.Lock()
-			defer p.lock.Unlock()
-
-			for len(p.pendingNotifications) == 0 {
-				// check if we're shutdown
-				select {
-				case <-stopCh:
-					return nil, true
-				default:
-				}
-				p.cond.Wait()
-			}
-
-			nt := p.pendingNotifications[0]
-			p.pendingNotifications = p.pendingNotifications[1:]
-			return nt, false
-		}
-
-		notification, stopped := blockingGet()
-		if stopped {
-			return
-		}
-
 		select {
 		case <-stopCh:
 			return
-		case p.nextCh <- notification:
+		case nextCh <- notification:
+			// Notification dispatched
+			if len(pendingNotifications) == 0 { // Nothing to pop
+				nextCh = nil
+				notification = nil
+			} else {
+				notification = pendingNotifications[0]
+				pendingNotifications[0] = nil
```

Comment:
What I have seen testing this in the go playground...

- Using `s = s[1:]`
```go
// create a slice of some random type T
s := []*T{x, y, z}
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 3, 3

// reset s to elements 1..n
s = s[1:]
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 2, 2

c := cap(s)
fmt.Printf("%v\n", s[0:c])
// prints out just 2 elements; the original first element is gone
```

- Using `s = append(s[:0], s[1:]...)
```go
// create a slice of some random type T
s := []*T{x, y, z}
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 3, 3

// use append to delete the first element
s = append(s[:0], s[1:]...)
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 2, 3

c := cap(s)
fmt.Printf("%v\n", s[0:c])
// prints out 3 elements: y, z, z
```

I think because we are specifically removing the first element using `s = s[1:]` we will be ok w.r.t. GC.